crates/ruff_python_formatter/src/statement/clause.rs

```diff
@@ -0,0 +1,350 @@
+            | ClauseHeader::With(_)
+            | ClauseHeader::OrElse(_) => last_child_end,
+
+            ClauseHeader::ExceptHandler(handler) => handler
+                .name
+                .as_ref()
+                .map(ruff_python_ast::Ranged::end)
+                .or(last_child_end),
+        };
+
+        let colon = colon_range(end.unwrap_or(keyword_range.end()), source)?;
+
+        Ok(TextRange::new(keyword_range.start(), colon.end()))
+    }
+
+    /// Visits the nodes in the case header.
+    pub(crate) fn visit<F>(self, visitor: &mut F)
+    where
+        F: FnMut(AnyNodeRef),
+    {
+        fn visit<'a, N, F>(node: N, visitor: &mut F)
+        where
+            N: Into<AnyNodeRef<'a>>,
+            F: FnMut(AnyNodeRef<'a>),
+        {
+            visitor(node.into());
+        }
+
+        match self {
+            ClauseHeader::Class(class) => {
+                if let Some(type_params) = &class.type_params {
+                    visit(type_params.as_ref(), visitor);
+                }
+
+                if let Some(arguments) = &class.arguments {
+                    visit(arguments.as_ref(), visitor);
+                }
+            }
+            ClauseHeader::Function(function) => {
+                if let Some(type_params) = function.type_params.as_ref() {
+                    visit(type_params, visitor);
+                }
+                visit(function.parameters.as_ref(), visitor);
+            }
+            ClauseHeader::If(if_stmt) => {
+                visit(if_stmt.test.as_ref(), visitor);
+            }
+            ClauseHeader::ElifElse(clause) => {
+                if let Some(test) = clause.test.as_ref() {
+                    visit(test, visitor);
+                }
+            }
+
+            ClauseHeader::ExceptHandler(handler) => {
+                if let Some(ty) = handler.type_.as_deref() {
+                    visit(ty, visitor);
+                }
+            }
+            ClauseHeader::Match(match_stmt) => {
+                visit(match_stmt.subject.as_ref(), visitor);
+            }
+            ClauseHeader::MatchCase(match_case) => {
+                visit(&match_case.pattern, visitor);
+
+                if let Some(guard) = match_case.guard.as_deref() {
+                    visit(guard, visitor);
+                }
+            }
+            ClauseHeader::For(for_stmt) => {
+                visit(for_stmt.target.as_ref(), visitor);
+                visit(for_stmt.iter.as_ref(), visitor);
+            }
+            ClauseHeader::While(while_stmt) => {
+                visit(while_stmt.test.as_ref(), visitor);
+            }
+            ClauseHeader::With(with_stmt) => {
+                for item in &with_stmt.items {
+                    visit(item, visitor);
+                }
+            }
+            ClauseHeader::Try(_) | ClauseHeader::TryFinally(_) | ClauseHeader::OrElse(_) => {}
+        }
+    }
+
+    fn keyword_range(self, source: &str) -> FormatResult<TextRange> {
```

Comment:
Nit: document that this returns the _first_ keyword in the clause (e.g., `async` in `async for`).